_posts/2023-08-07-biquery-partitioning.md

```diff
@@ -0,0 +1,185 @@
+---
+title: "BigQuery partitioning - by time-unit column or by ingestion time"
+date: 2023-08-07
+tags: [GCP, BigQuery]
+header:
+    overlay_image: /assets/images/2022-03-24-gcp-dataproc-spark-tuning/nana-smirnova-IEiAmhXehwE-unsplash.webp
+    caption: "[Unsplash](https://unsplash.com/@nanasmirnova)"
+---
+
+What's the best way to partition time-series data in BigQuery?
+By time-unit column or by ingestion time? Daily or hourly?
+It depends, keep reading to learn trade-offs, pitfalls, and other traps.
+
+## Time-series or aggregates
+
+For time-series with data points with minute, second or millisecond precision use time-unit column partitioning.
+You will get convenient querying like:
+
+```sql
+SELECT ts, temperate, pressure
+FROM weather_stations
+WHERE
+    ts BETWEEN TIMESTAMP("2008-12-25 15:30:00")
+        AND TIMESTAMP("2008-12-25 15:35:00")
+```
+
+With ingestion time partitioning, you have to specify extra predicate for partition pruning:
+
+```sql
+SELECT ts, temperate, pressure
+FROM weather_stations
+WHERE
+    ts BETWEEN TIMESTAMP("2008-12-25 15:30:00")
+        AND TIMESTAMP("2008-12-25 15:35:00")
+    AND _PARTITIONTIME = DATE("2008-12-15")
+```
+
+For aggregates, when you typically don't need an exact point in time, you can use ingestion time partitioning.
+You don't need to specify the time column explicitly in a table schema.
+
+```sql
+SELECT _PARTITIONTIME AS dt, page_views, unique_visitors
+FROM ecommerce_sessions
+WHERE
+    _PARTITIONTIME = DATE("2008-12-15")
+```
+
+## Retention
+
+In my projects, the majority of tables require at least 1--3 years of history.
+With a limit of 4000 partitions per BigQuery table, it requires at least daily partitioning.
+Tables with 3 years of retention use `3 * 365 = 1095` daily partitions, below limit.
+Tables with hourly partitions keep up to only `4000 / 24 = 166 days and 8 hours` of data.
+
+For tables with more than 10 years of history I would consider another storage than BigQuery.
+[Export](https://cloud.google.com/bigquery/docs/exporting-data) partitions with data older than 10 years
+to [Google Cloud Storage](https://cloud.google.com/storage/docs) bucket using columnar format like
+[Apache Parquet](https://parquet.apache.org/docs/).
+Create BigQuery [external table](https://cloud.google.com/bigquery/docs/external-data-cloud-storage) on the bucket
+using Hive partitioning scheme for querying.
```

Comment:
Dobrze, że nie dodałeś, że najlepiej w klasie Archive, bo za podnoszenie danych się na razie i tak nie płaci ;p

Reply:
To już zdecydowanie na inny post